be/src/io/s3_output_stream.cpp

```diff
@@ -0,0 +1,151 @@
+
+#include "io/s3_output_stream.h"
+
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/CompleteMultipartUploadRequest.h>
+#include <aws/s3/model/CreateMultipartUploadRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/s3/model/UploadPartRequest.h>
+#include <fmt/format.h>
+
+#include "common/logging.h"
+
+namespace starrocks::io {
+
+S3OutputStream::S3OutputStream(std::shared_ptr<Aws::S3::S3Client> client, std::string bucket, std::string object,
+                               int64_t max_single_part_size, int64_t min_upload_part_size)
+        : _client(std::move(client)),
+          _bucket(std::move(bucket)),
+          _object(std::move(object)),
+          _max_single_part_size(max_single_part_size),
+          _min_upload_part_size(min_upload_part_size),
+          _buffer(),
+          _upload_id(),
+          _etags() {
+    CHECK(_client != nullptr);
+}
+
+Status S3OutputStream::write(const void* data, int64_t size) {
+    _buffer.append(static_cast<const char*>(data), size);
+    if (_upload_id.empty() && _buffer.size() > _max_single_part_size) {
+        RETURN_IF_ERROR(create_multipart_upload());
+        DCHECK(!_upload_id.empty());
+    }
+    if (!_upload_id.empty() && _buffer.size() >= _min_upload_part_size) {
+        RETURN_IF_ERROR(multipart_upload());
+        _buffer.clear();
+    }
+    return Status::OK();
+}
+
+Status S3OutputStream::write_aliased(const void* data, int64_t size) {
+    return Status::NotSupported("S3OutputStream::write_aliased");
+}
+
+Status S3OutputStream::skip(int64_t count) {
+    _buffer.resize(_buffer.size() + count);
+    // Calling `write()` will trigger the uploading checks.
+    return S3OutputStream::write("", 0);
+}
+
+StatusOr<OutputStream::Buffer> S3OutputStream::get_direct_buffer() {
+    return Buffer();
+}
+
+StatusOr<OutputStream::Position> S3OutputStream::get_direct_buffer_and_advance(int64_t size) {
+    auto old_size = _buffer.size();
+    _buffer.resize(old_size + size);
+    return reinterpret_cast<uint8_t*>(_buffer.data()) + old_size;
+}
+
+Status S3OutputStream::close() {
+    if (_upload_id.empty()) {
+        RETURN_IF_ERROR(singlepart_upload());
+    } else {
+        RETURN_IF_ERROR(multipart_upload());
+        RETURN_IF_ERROR(complete_multipart_upload());
+    }
+    _client = nullptr;
+    return Status::OK();
+}
+
+Status S3OutputStream::create_multipart_upload() {
+    Aws::S3::Model::CreateMultipartUploadRequest req;
+    req.SetBucket(_bucket);
+    req.SetKey(_object);
+    Aws::S3::Model::CreateMultipartUploadOutcome outcome = _client->CreateMultipartUpload(req);
+    if (outcome.IsSuccess()) {
+        _upload_id = outcome.GetResult().GetUploadId();
+        return Status::OK();
+    }
+    return Status::IOError(fmt::format("S3: Fail to create multipart upload: {}", outcome.GetError().GetMessage()));
+}
+
+Status S3OutputStream::singlepart_upload() {
+    VLOG(12) << "Uploading s3://" << _bucket << "/" << _object << " vis singlepart upload";
```

Comment:
s/vis/via

Reply:
Fixed, PTAL

Reply:
what does 12 mean in VLOG(12)?

Reply:
```
With VLOG, the lower the verbose level, the more likely messages are to be logged. For example, if --v==1, VLOG(1) will log, but VLOG(2) will not log. This is opposite of the severity level, where INFO is 0, and ERROR is 2. --minloglevel of 1 will log WARNING and above. Though you can specify any integers for both VLOG macro and --v flag, the common values for them are small positive integers. For example, if you write VLOG(0), you should specify --v=-1 or lower to silence it. This is less useful since we may not want verbose logs by default in most cases. The VLOG macros always log at the INFO log level (when they log at all).
```
https://rpg.ifi.uzh.ch/docs/glog.html